### ISSUE TITLE
rustc nightly build crashes when generating doc for crossbeam v0.8.1

### ISSUE THREAD

**Issue by vmingchen** [rust-lang/rust#95793]
The crash happened during Glommio's Github "doc-nightly" check. It happened consistently for three times:

1. https://github.com/DataDog/glommio/runs/5870536608?check_suite_focus=true
2. https://github.com/DataDog/glommio/runs/5860399018?check_suite_focus=true
3. https://github.com/DataDog/glommio/runs/5860206267?check_suite_focus=true

### Code

```Rust
<code>
```

### Meta
This is a bug about nightly build.

```
note: rustc 1.62.0-nightly (8f36334ca 2022-04-06) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type lib
```

### Error output

```
thread 'rustc' panicked at 'no entry found for key', src/librustdoc/passes/collect_intra_doc_links.rs:977:16

```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
thread 'rustc' panicked at 'no entry found for key', src/librustdoc/passes/collect_intra_doc_links.rs:977:16
stack backtrace:
   0:     0x7f3f418a9cbd - std::backtrace_rs::backtrace::libunwind::trace::h39a05c79[153](https://github.com/DataDog/glommio/runs/5870536608?check_suite_focus=true#step:5:153)cf59e
                               at /rustc/8f36334ca939a67cce3f37f24953ff6f2d3f3d33/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7f3f418a9cbd - std::backtrace_rs::backtrace::trace_unsynchronized::hb601deac92d79a2c
                               at /rustc/8f36334ca939a67cce3f37f24953ff6f2d3f3d33/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f3f418a9cbd - std::sys_common::backtrace::_print_fmt::hb75d966e9ab3e621
                               at /rustc/8f36334ca939a67cce3f37f24953ff6f2d3f3d33/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x7f3f418a9cbd - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hd34f84efb52430f5
                               at /rustc/8f36334ca939a67cce3f37f24953ff6f2d3f3d33/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x7f3f41903acc - core::fmt::write::hc743d95c91409805
                               at /rustc/8f36334ca939a67cce3f37f24953ff6f2d3f3d33/library/core/src/fmt/mod.rs:1194:17
   5:     0x7f3f4189b4c1 - std::io::Write::write_fmt::h92fccc03b1282e9e
                               at /rustc/8f36334ca939a67cce3f37f24953ff6f2d3f3d33/library/std/src/io/mod.rs:1655:15
   6:     0x7f3f418acbe5 - std::sys_common::backtrace::_print::h63a8f68e57f87126
                               at /rustc/8f36334ca939a67cce3f37f24953ff6f2d3f3d33/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x7f3f418acbe5 - std::sys_common::backtrace::print::hadfeebfadb51d1c9
                               at /rustc/8f36334ca939a67cce3f37f24953ff6f2d3f3d33/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x7f3f418acbe5 - std::panicking::default_hook::{{closure}}::h107f9f9f118a98ac
                               at /rustc/8f36334ca939a67cce3f37f24953ff6f2d3f3d33/library/std/src/panicking.rs:295:22
   9:     0x7f3f418ac859 - std::panicking::default_hook::h22bdc59447aa0316
                               at /rustc/8f36334ca939a67cce3f37f24953ff6f2d3f3d33/library/std/src/panicking.rs:314:9
  10:     0x7f3f4204b4c1 - rustc_driver[20c8da45edb322c4]::DEFAULT_HOOK::{closure#0}::{closure#0}
  11:     0x7f3f418ad380 - std::panicking::rust_panic_with_hook::hffa936ba366a6e55
                               at /rustc/8f36334ca939a67cce3f37f24953ff6f2d3f3d33/library/std/src/panicking.rs:702:17
  12:     0x7f3f418ad1b7 - std::panicking::begin_panic_handler::{{closure}}::hc69b957e62852aab
                               at /rustc/8f36334ca939a67cce3f37f24953ff6f2d3f3d33/library/std/src/panicking.rs:588:13
  13:     0x7f3f418aa174 - std::sys_common::backtrace::__rust_end_short_backtrace::hd4ba2e5dd738ece1
                               at /rustc/8f36334ca939a67cce3f37f24953ff6f2d3f3d33/library/std/src/sys_common/backtrace.rs:138:18
  14:     0x7f3f418acee9 - rust_begin_unwind
                               at /rustc/8f36334ca939a67cce3f37f24953ff6f2d3f3d33/library/std/src/panicking.rs:584:5
  15:     0x7f3f418715a3 - core::panicking::panic_fmt::h1c45726dd4016180
                               at /rustc/8f36334ca939a67cce3f37f24953ff6f2d3f3d33/library/core/src/panicking.rs:143:14
  16:     0x7f3f419008e1 - core::panicking::panic_display::h408675489cc9369d
                               at /rustc/8f36334ca939a67cce3f37f24953ff6f2d3f3d33/library/core/src/panicking.rs:72:5
  17:     0x7f3f4190088b - core::panicking::panic_str::h503ed8e0c88cc2af
                               at /rustc/8f36334ca939a67cce3f37f24953ff6f2d3f3d33/library/core/src/panicking.rs:56:5
  18:     0x7f3f41871416 - core::option::expect_failed::h6bc6bc22da6ab201
                               at /rustc/8f36334ca939a67cce3f37f24953ff6f2d3f3d33/library/core/src/option.rs:1874:5
  19:     0x55c9bd296af9 - rustdoc[a5742160089a3f9a]::passes::collect_intra_doc_links::resolve_associated_trait_item
  20:     0x55c9bd29530e - <rustdoc[a5742160089a3f9a]::passes::collect_intra_doc_links::LinkCollector>::resolve_associated_item
  21:     0x55c9bd2924f5 - <rustdoc[a5742160089a3f9a]::passes::collect_intra_doc_links::LinkCollector>::resolve
  22:     0x55c9bd298605 - <rustdoc[a5742160089a3f9a]::passes::collect_intra_doc_links::LinkCollector as rustdoc[a5742160089a3f9a]::visit::DocVisitor>::visit_item
  23:     0x55c9bd2a4b3a - <rustdoc[a5742160089a3f9a]::passes::collect_intra_doc_links::LinkCollector as rustdoc[a5742160089a3f9a]::visit::DocVisitor>::visit_inner_recur
  24:     0x55c9bd29b894 - <rustdoc[a5742160089a3f9a]::passes::collect_intra_doc_links::LinkCollector as rustdoc[a5742160089a3f9a]::visit::DocVisitor>::visit_item
  25:     0x55c9bd290cec - rustdoc[a5742160089a3f9a]::passes::collect_intra_doc_links::collect_intra_doc_links
  26:     0x55c9bd2aa9d9 - <rustc_session[6907ade8b72e179b]::session::Session>::time::<(rustdoc[a5742160089a3f9a]::clean::types::Crate, rustdoc[a5742160089a3f9a]::config::RenderOptions, rustdoc[a5742160089a3f9a]::formats::cache::Cache), rustdoc[a5742160089a3f9a]::main_options::{closure#0}::{closure#0}::{closure#1}::{closure#0}>
  27:     0x55c9bd2f8a16 - <rustc_interface[d43dd518ab95666e]::passes::QueryContext>::enter::<rustdoc[a5742160089a3f9a]::main_options::{closure#0}::{closure#0}::{closure#1}, core[c075b761217c3c62]::result::Result<(), rustc_errors[88efdcb8dda3d889]::ErrorGuaranteed>>
  28:     0x55c9bd239ea9 - <rustc_interface[d43dd518ab95666e]::interface::Compiler>::enter::<rustdoc[a5742160089a3f9a]::main_options::{closure#0}::{closure#0}, core[c075b761217c3c62]::result::Result<(), rustc_errors[88efdcb8dda3d889]::ErrorGuaranteed>>
  29:     0x55c9bd39b896 - rustc_span[c12cfe48d2ae16cd]::with_source_map::<core[c075b761217c3c62]::result::Result<(), rustc_errors[88efdcb8dda3d889]::ErrorGuaranteed>, rustc_interface[d43dd518ab95666e]::interface::create_compiler_and_run<core[c075b761217c3c62]::result::Result<(), rustc_errors[88efdcb8dda3d889]::ErrorGuaranteed>, rustdoc[a5742160089a3f9a]::main_options::{closure#0}>::{closure#1}>
  30:     0x55c9bd271074 - rustc_interface[d43dd518ab95666e]::interface::create_compiler_and_run::<core[c075b761217c3c62]::result::Result<(), rustc_errors[88efdcb8dda3d889]::ErrorGuaranteed>, rustdoc[a5742160089a3f9a]::main_options::{closure#0}>
  31:     0x55c9bd0ced25 - <scoped_tls[556e710d4500340c]::ScopedKey<rustc_span[c12cfe48d2ae16cd]::SessionGlobals>>::set::<rustdoc[a5742160089a3f9a]::main_args::{closure#0}, core[c075b761217c3c62]::result::Result<(), rustc_errors[88efdcb8dda3d889]::ErrorGuaranteed>>
  32:     0x55c9bd220b6f - std[928ba903c221ea94]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[d43dd518ab95666e]::util::run_in_thread_pool_with_globals<rustdoc[a5742160089a3f9a]::main_args::{closure#0}, core[c075b761217c3c62]::result::Result<(), rustc_errors[88efdcb8dda3d889]::ErrorGuaranteed>>::{closure#0}, core[c075b761217c3c62]::result::Result<(), rustc_errors[88efdcb8dda3d889]::ErrorGuaranteed>>
  33:     0x55c9bd31a219 - <<std[928ba903c221ea94]::thread::Builder>::spawn_unchecked_<rustc_interface[d43dd518ab95666e]::util::run_in_thread_pool_with_globals<rustdoc[a5742160089a3f9a]::main_args::{closure#0}, core[c075b761217c3c62]::result::Result<(), rustc_errors[88efdcb8dda3d889]::ErrorGuaranteed>>::{closure#0}, core[c075b761217c3c62]::result::Result<(), rustc_errors[88efdcb8dda3d889]::ErrorGuaranteed>>::{closure#1} as core[c075b761217c3c62]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  34:     0x7f3f418b7453 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::ha611c1ce8a641400
                               at /rustc/8f36334ca939a67cce3f37f24953ff6f2d3f3d33/library/alloc/src/boxed.rs:1861:9
  35:     0x7f3f418b7453 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h50ba8bc6f1945fe3
                               at /rustc/8f36334ca939a67cce3f37f24953ff6f2d3f3d33/library/alloc/src/boxed.rs:1861:9
  36:     0x7f3f418b7453 - std::sys::unix::thread::Thread::new::thread_start::h4193a3a6c78b3496
                               at /rustc/8f36334ca939a67cce3f37f24953ff6f2d3f3d33/library/std/src/sys/unix/thread.rs:108:17
  37:     0x7f3f417e2609 - start_thread
  38:     0x7f3f4[158](https://github.com/DataDog/glommio/runs/5870536608?check_suite_focus=true#step:5:158)b[163](https://github.com/DataDog/glommio/runs/5870536608?check_suite_focus=true#step:5:163) - clone
  39:                0x0 - <unknown>
```

</p>
</details>

**Comment by mcvicuna** [comment]
This was fixed in commit dd38eea722c1f6f970a314435533e91cc2c14ffb of nightly-2022-04-08 and the test triggers the same ICE in nightly-022-04-07:
```
thread 'rustc' panicked at 'no entry found for key', src/librustdoc/passes/collect_intra_doc_links.rs:977:16
```

**Comment by petrochenkov** [comment]
Closing as fixed.
